README.md
# KeePassium Password Manager

[KeePassium](https://keepassium.com) is a KeePass-compatible password manager for iOS. It offers automatic database synchronization, respect to privacy and premium user experience.

KeePassium is a commercial open source app. The "commercial" part gives you a well-maintained app with premium support. The "open source" part gives you the transparency expected from a password manager: you can personally audit the code and build the app you can trust.

## Contents

- [Screenshots](#screenshots)
- [Features](#features)
	- [Automatic sync](#automatic-sync)
	- [Password AutoFill](#password-autofill)
	- [App and Data Protection](#app-and-data-protection)
- [Installation](#installation)
- [Is it free?](#is-it-free)
- [How to Contribute](#how-to-contribute)
- [Support](#support)
- [Author and Credits](#author-and-credits)
- [License](#license)

## Screenshots

![Database Unlock](https://keepassium.com/img/github/unlock-db_w250.png) ![View Group](https://keepassium.com/img/github/view-group_w250.png) ![View Entry](https://keepassium.com/img/github/view-entry_w250.png)

## Features

* [Automatic database synchronization](#automatic-sync) with zero setup
	- Integrates with the system, does not ask for your storage credentials
	- Works with iCloud Drive, Dropbox, OneDrive, Google Drive, Box, Nextcloud, SFTP, and probably more.
* [Password AutoFill (iOS 12+)](#password-autofill) — also with synchronization.
* [App and data protection](#app-and-data-protection):
	- Biometric (Face ID / Touch ID) and passcode-based protection.
	- Customizable timeouts for app, database and clipboard.
	- Database names and app settings are also protected.
* Read/write support for all KeePass formats:
	- `kdbx4` (KeePass 2.35+), `kdbx3` (KeePass 2.x) and `kdb` (KeePass 1.x)
	- ChaCha20, Argon2, AES, Salsa20, Twofish algorithms
* Easy switching between multiple databases
* And more:
	- Viewing TOTP codes (both RFC-6238 and Steam TOTP)
	- File attachments
	- Custom fields
	- Custom icons
	- No ads

### Automatic sync

KeePassium relies on OS-provided integration with storage providers, and supports most providers that appear in iOS Files app.

* __Full support__ (files are automatically downloaded and uploaded when changed): iCloud Drive, Box, [Dropbox](https://keepassium.com/articles/sync-ios-keepass-with-dropbox/), [Google Drive](https://keepassium.com/articles/sync-ios-keepass-with-google-drive/), [OneDrive](https://keepassium.com/articles/sync-ios-keepass-with-onedrive/), [Resilio Sync](https://keepassium.com/articles/sync-ios-keepass-with-resilio-sync/), [Nextcloud](https://keepassium.com/articles/sync-ios-keepass-with-nextcloud/) (with hiccups), [SFTP](https://keepassium.com/articles/sync-ios-keepass-with-ftp-sftp/) / [WebDAV](https://keepassium.com/articles/sync-ios-keepass-with-webdav/) (via third-party apps)
* __Limited support__ (no background sync, files have to be imported and exported [manually](https://keepassium.com/articles/how-to-sync-database-manually/)): Mega, Cryptomator (it supports [only import/export](https://github.com/cryptomator/cryptomator-ios/issues/98#issuecomment-402446441) operations)
* __Not tested yet__: ownCloud, Tresorit

If automatic sync does not work for you, please make sure the cloud provider app can work in background (device settings — Dropbox/OneDrive/... — Background App Refresh = Enabled). 


### Password AutoFill

To fill your passwords easily and quickly, [enable AutoFill feature](https://keepassium.com/articles/how-to-setup-autofill/): device settings — Passwords & Accounts — AutoFill Passwords — select KeePassium in the list.

### App and Data Protection

KeePassium has multi-layer protection from unauthorized access.

- __App Lock__ protects the app itself, by covering any in-app screens. This way, only you can see the unlocked database, the list of databases, or change app settings. 
- __Database Lock__ closes all opened databases after a timeout, or after a failed AppLock attempt. It also removes any remembered master keys from keychain.
- __Encryption__ protects the contents of database files. The app relies on [CommonCrypto](https://opensource.apple.com/source/CommonCrypto/) library (for AES and SHA) and time-proven reference implementations of crypto algorithms (see [Credits](#author-and-credits)).
 
By default, AppLock requires a passcode (of any complexity). For convenience, you can configure the app to use Face ID/Touch ID instead.


## Installation

KeePassium is [available on the App Store](https://apps.apple.com/us/app/id1435127111). Current beta can be installed via [TestFlight](https://testflight.apple.com/join/y8R6iLlK). 

To build your own binary, download the project and open it in Xcode 10.2 (or above). All the dependencies are already included.


## Is it free?

[Yes](https://keepassium.com/articles/is-keepassium-free/), KeePassium can be used free of charge (gratis) and is also [free as in speech](#license) (libre). An optional [premium subscription](https://keepassium.com/articles/why-upgrade-to-premium/) enables a few convenience and time-saving features, and supports the project.

**But I can just...** Yes, you can take the source code and build your personal premium version for free (gratis). Feel free to do so, but please be nice and don't request premium support for personal builds. Also, if you delegate this to a freelancer, please make sure you can trust them. 


## How to Contribute

Code contributions cannot be accepted due to [formalities](#license), but feel free to:

- [Report bugs, suggest new features](https://github.com/keepassium/KeePassium/issues)
- [Contribute to translation](https://github.com/keepassium/KeePassium-L10n)
- [Write an AppStore review](https://apps.apple.com/us/app/id1435127111)
- Consider buying a premium subscription :)

## Support

- Bug reports: [GitHub Issues](https://github.com/keepassium/KeePassium/issues)
- Discussion: [/r/KeePassium](https://reddit.com/r/KeePassium)
- Updates: [KeePassium Blog](https://keepassium.com/blog/) and [Twitter](https://twitter.com/keepassium)


## Author and Credits

KeePassium is created and maintained by [Andrei Popleteev](http://popleteev.com) (also responsible for [KeePassB for BlackBerry 10](https://github.com/anmipo/keepassb)).

The project would not be the same without some third-party components:

* Graphics:
	- [Feather icons](https://feathericons.com) by Cole Bemis (MIT licence) 	
	- [Ionicons](http://ionicons.com) by Ionic (MIT license)
	- [Linecons](https://designmodo.com/linecons-free/) by Andrian Valeanu (CC-BY-ND 3.0 license)
	- [Fancy deboss pattern](http://subtlepatterns.com) by Daniel Beaton (CC-BY-3.0 when downloaded)
	- [System settings icon](https://www.iconfinder.com/icons/2697651/apple_configuration_control_gear_preferences_setting_settings_icon) by Vicons Design (CC-BY 3.0 license)
	- [Bold outline icons](https://roundicons.com/boldicons-outline-icons-set/) by Round Icons ([commercial license](https://roundicons.com/usage-license/)). They cannot be shared and have been replaced with standard KeePass icons from the [Nuvola icon set](https://en.wikipedia.org/wiki/Nuvola) by David Vignoni (LGPL v2.1).
* Code:
	- [AEXML](https://github.com/tadija/AEXML) by Marko Tadić (MIT license) 	
	- Rijndael implementation by Szymon Stefanek (public domain)
	- [Argon2](https://github.com/P-H-C/phc-winner-argon2) by Daniel Dinu, Dmitry Khovratovich, Jean-Philippe Aumasson, and Samuel Neves (CC0 license)
	- [ChaCha20 & Salsa20](https://cr.yp.to/salsa20.html) implementation by D. J. Bernstein (public domain)
	- [Twofish](http://www.cartotype.com/downloads/twofish/twofish.cpp) implementation by Niels Ferguson (custom very permissive license).
	- [Yubico Mobile iOS SDK](https://github.com/Yubico/yubikit-ios) by Yubico AB (Apache 2.0 license)
	- [TPInAppReceipt](https://github.com/tikhop/TPInAppReceipt) by Pavel Tikhonenko (MIT license)
	- [GzipSwift](https://github.com/1024jp/GzipSwift) by 1024jp (MIT license)
	- [Base32 for Swift](https://github.com/norio-nomura/Base32) by Norio Nomura (MIT license)
	- [KeyboardLayoutConstraint](https://github.com/MengTo/Spring/blob/master/Spring/KeyboardLayoutConstraint.swift) by James Tang (MIT license)

To avoid backdoors in third-party code, it has been checked by the main developer. Verified files are directly included in the project — we don't want any surprises sneaking via package managers.

## License

KeePassium Password Manager

Copyright ©2018–2019 [Andrei Popleteev](http://popleteev.com).

KeePassium is a commercial open-source app, available under the  [GPLv3 license](https://choosealicense.com/licenses/gpl-3.0/). Our intention is to provide the maximal possible transparency: you can personally audit the code and build your own binary. 

While derivative works (forks) are explicitly allowed by the GPL, please don't submit them to AppStore. Due to a conflict between GPL and AppStore terms of service, GPL-licensed apps are [banned](https://www.fsf.org/blogs/licensing/more-about-the-app-store-gpl-enforcement) from AppStore. For the same reason, KeePassium [cannot accept code contributions](https://apple.stackexchange.com/questions/6109/is-it-possible-to-have-gpl-software-in-the-mac-app-store).

For commercial licensing or custom modifications, please [contact us](info@keepassium.com).


CHANGELOG.md
#CHANGELOG

## [1.19.76] - 2020-10-03

### Changed

- Enabled express unlock in the free version.

### Fixed

- Possible database corruption on save in v1.18 (fixes #130) [thanks, everyone]
- URL fields not recognized on iOS 14 if default browser is not Safari (fixes #129) [thanks, Ivo]


## [1.18.75] - 2020-10-01

### Added

- [Premium] Express unlock using the stored decryption key. This also allows opening YubiKey-protected databases in AutoFill.

### Changed

- Free version won't count broken databases towards the free DB limit
- Added detection of Stratospherix FileBrowser file provider
- Refined local storage icons

### Fixed

- Slow decryption of databases (fixes #128) [thanks, everyone]
- Preserve DB settings on DB deletion, if another DB uses them [thanks, Tim]
- Crash when unselecting a YubiKey in AutoFill
- Detection of file access permission error
- Delete .latest backups when manually deleting all backups via settings


## [1.17.74] - 2020-09-19

### Changed
 
- (iOS 14) Suggesting to re-add the database if it seems missing.
- Keeping the .latest backup file regardless of its age
- Refined CallerID in AutoFill to make it easier to copy [thanks, Adam]
- Showing a wait animation when deleting many backup files
- Increased the max length of generated passwords to 100 [thanks, u/ReevaluateAdNauseam]

### Fixed

- Recognizing storage providers on iOS 14
- Freezing when there are hundreds of backup files (related #109) [thanks, everyone]
- Opening files with unregistered extensions [thanks, Kevin]
- Displaying multi-line Caller ID URLs in AutoFill [thanks, Adam]
- App icon picker UI on iOS 12 (closes #123) [thanks, dragonblitz10]
- Sometimes backup files were shown as external ones
- Opening kdbx4 files with UInt32 transformRounds (keeweb/keeweb#1598) [thanks, Mitchell]
- Hopefully fixed: text missing in entry viewer [thanks, Lionel and Jean-Marc]
- Splash screen background color in dark mode


## [1.16.73] - 2020-08-31

### Added

- Change entry text size with a zoom gesture [thanks, Rick and Thomas]

### Changed

- Refined: preserving the expanded/collapsed state of the Notes field [thanks, Arjan and Thomas]
- Improved Export and Trash icons [thanks, Ivo]
- Updated translations [thanks, everyone]

### Fixed

- Help article text was black in dark mode [thanks, Christian]


## [1.15.72] - 2020-08-28

### Fixed

- Replaced "Try it free" with "Upgrade now" to address AppStore reviewer's comment


## [1.15.71] - 2020-08-28

### Fixed

- Moved trial conditions to the purchase button itself (to address AppStore reviewer's comment)


## [1.15.70] - 2020-08-27

### Added

- Perpetual fallback license for subscriptions (a.k.a. "rent-to-own license")
- App history screen (What's New)

### Changed

- Improved search with diacritics: insensitive to diacritics, unless you use diacritics in the query (closes #118) [thanks, hunhejj]
- Improved haptic feedback in AutoFill
- Updated translations [thanks, everyone]

### Fixed 

- Replacing broken references to external files failed sometimes [thanks, Paul]
- Relative project file paths on GitHub (thanks, @mj)


## [1.14.69] - 2020-08-11

### Added

- Special icon for databases in Trash (Recently Deleted) folder

### Fixed

- Crash after updating the app (iPad) [thanks, everyone]
- Crash when adding an entry attachment	[thanks, Sophie]
- File refresh spinner were invisible in dark mode
- Some icons in the settings


## [1.14.68] - 2020-08-04

### Added

- Customizable app icon
- Offline fallback option: KeePassium maintains a backup copy of the last loaded/saved version of database

### Changed

- Default app icon changed to Atom Blue (freemium) and Atom Black (Pro)
- By default, backups are now kept for 2 months instead of forever
- More detailed file location description for internal files
- Code: switched to KeePassium's own, checked fork of TPInAppReceipt
- AppLock setup reminder can be dismissed [thanks, Dan]
- Minor UI refinements here and there
- Updated DE/FR/JA/NL/RU translations [thanks, everyone]

### Removed

- Xcode11GM_LocalizedLabel, an obsolete workaround for #60

### Fixed

- Excessive caching of file attributes (size, timestamps)
- After transferring to a new device, AppLock did not always accept the correct passcode [thanks, Kirsten]
- Add note that Universal Clipboard timeout on external devices is fixed at 2 minutes [thanks, Christian]
- Backup files with non-standard extensions were not shown
- Tapping the "Show backup files" could change the file sort order instead
- Sporadic crash immediately after loading a database
- AutoFill Setup smallprint always appeared in English
- Return key behavior when changing DB master key
- Hardware key picker is localized now


## [1.14.67] - 2020-07-25

### Changed

- More human-friendly messages about missing file provider
- Replaced problematic NSFileCoordinator calls with UIDocument ones
- Refined purchase dialog UI on iPad
- Updated Japanese translation [thanks, miyar520]

### Fixed

- Fixed persistent timeouts when using Google Drive, DS file and some other clouds. [thanks, everyone]
- Fixed freezing when there are many backup files (fixes #114) [thanks, Paul]
- Added missing credits for Yubico Mobile SDK and TPInAppReceipt library


## [1.14.66] - 2020-07-18

### Added

- "Exclude from iCloud/iTunes backup" option for local files (closes #97) [thanks, everyone]
- Free trial period for premium subscriptions

### Changed

- Accepting databases with any file extension (closes #113) [thanks, everyone]
- Removed premium upgrade button from DB unlock screen
- Upgraded to Swift 5


## [1.14.65] - 2020-07-14

### Added

- New "Clear master keys on database timeout" switch to close-but-not-lock databases in multi-device scenarios [thanks, Niklas]
- New UI for premium upgrade

### Changed

- Increased file timeout from 5 to 15 seconds [thanks, everyone]
- More human-friendly error messages when cloud provider does not respond
- Updated Japanese and Russian translation [thanks, miyar520]

KeePassium/general/AboutVC.swift
//  KeePassium Password Manager
//  Copyright © 2018–2019 Andrei Popleteev <info@keepassium.com>
// 
//  This program is free software: you can redistribute it and/or modify it
//  under the terms of the GNU General Public License version 3 as published
//  by the Free Software Foundation: https://www.gnu.org/licenses/).
//  For commercial licensing, please contact the author.

import UIKit
import KeePassiumLib

class AboutVC: UITableViewController {
    @IBOutlet weak var contactSupportCell: UITableViewCell!
    @IBOutlet weak var writeReviewCell: UITableViewCell!
    @IBOutlet weak var versionLabel: UILabel!
    
    let cellTagToURL: [Int: String] = [
        10: "https://github.com/keepassium/KeePassium-L10n",
        20: "https://keepass.info",
        30: "https://feathericons.com",
        40: "http://ionicons.com",
        50: "https://designmodo.com/linecons-free/",
        55: "https://icons8.com/paid-license-99",
        60: "http://subtlepatterns.com",
        70: "http://vicons.superatic.com",
        80: "https://github.com/tadija/AEXML",
        90: "",
        100: "https://github.com/P-H-C/phc-winner-argon2",
        110: "https://cr.yp.to/salsa20.html",
        120: "http://www.cartotype.com/downloads/twofish/twofish.cpp",
        125: "https://github.com/tikhop/TPInAppReceipt",
        130: "https://github.com/1024jp/GzipSwift",
        140: "https://github.com/norio-nomura/Base32",
        150: "https://github.com/MengTo/Spring/blob/master/Spring/KeyboardLayoutConstraint.swift",
    ]
    
    static func make() -> UIViewController {
        let vc = AboutVC.instantiateFromStoryboard()
        return vc
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        tableView.rowHeight = UITableView.automaticDimension
        tableView.estimatedRowHeight = 44
        
        if Settings.current.isTestEnvironment {
            versionLabel.text = "v\(AppInfo.version).\(AppInfo.build) beta"
        } else {
            versionLabel.text = "v\(AppInfo.version).\(AppInfo.build)"
        }
    }
    
    override func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        if section == 0 {
            return 0.1
        } else {
            return super.tableView(tableView, heightForHeaderInSection: section)
        }
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        guard let selectedCell = tableView.cellForRow(at: indexPath) else { return }
        switch selectedCell {
        case contactSupportCell:
            SupportEmailComposer.show(subject: .supportRequest, parent: self)
        case writeReviewCell:
            AppStoreHelper.writeReview()
        default:
            if let urlString = cellTagToURL[selectedCell.tag], let url = URL(string: urlString) {
                AppGroup.applicationShared?.open(url, options: [:], completionHandler: nil)
            }
        } 
    }
}
